workspaces/arborist/lib/query-selector-all.js

```diff
@@ -0,0 +1,523 @@
+const { parser, arrayDelimiter } = require('@npmcli/query')
+const localeCompare = require('@isaacs/string-locale-compare')('en')
+const npa = require('npm-package-arg')
+const minimatch = require('minimatch')
+const semver = require('semver')
+
+// handle results for parsed query asts, results are stored in a map that
+// has a key that points to each ast selector node and stores the resulting
+// array of arborist nodes as its value, that is essential to how we handle
+// multiple query selectors, e.g: `#a, #b, #c` <- 3 diff ast selector nodes
+class Results {
+  #results = null
+  #currentAstSelector = null
+
+  constructor (rootAstNode) {
+    this.#results = new Map()
+    this.#currentAstSelector = rootAstNode.nodes[0]
+  }
+
+  /* eslint-disable-next-line accessor-pairs */
+  set currentAstSelector (value) {
+    this.#currentAstSelector = value
+  }
+
+  get currentResult () {
+    return this.#results.get(this.#currentAstSelector)
+  }
+
+  set currentResult (value) {
+    this.#results.set(this.#currentAstSelector, value)
+  }
+
+  // when collecting results to a root astNode, we traverse the list of
+  // child selector nodes and collect all of their resulting arborist nodes
+  // into a single/flat Set of items, this ensures we also deduplicate items
+  collect (rootAstNode) {
+    const acc = new Set()
+    for (const n of rootAstNode.nodes) {
+      for (const node of this.#results.get(n)) {
+        acc.add(node)
+      }
+    }
+    return acc
+  }
+}
+
+const parentCache = new Map()
+const retrieveNodesFromParsedAst = async ({
+  initialItems,
+  inventory,
+  rootAstNode,
+  targetNode,
+}) => {
+  if (!rootAstNode.nodes) {
+    return new Set()
+  }
+
+  const ArboristNode = targetNode.constructor
+
+  const results = new Results(rootAstNode)
+  let currentAstNode = rootAstNode
+  let pendingCombinator = null
+
+  results.currentResult = initialItems
+
+  // maps containing the logic to parse each of the supported css selectors
+  const attributeOperatorsMap = new Map(Object.entries({
+    '' ({ attribute, value, pkg }) {
+      return Boolean(pkg[attribute])
+    },
+    '=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '') === value
+    },
+    '~=' ({ attribute, value, pkg }) {
+      return (String(pkg[attribute] || '').match(/\w+/g) || []).includes(value)
+    },
+    '*=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '').indexOf(value) > -1
+    },
+    '|=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '').split('-')[0] === value
+    },
+    '^=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '').startsWith(value)
+    },
+    '$=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '').endsWith(value)
+    },
+  }))
+  const classesMap = new Map(Object.entries({
+    '.prod' (prevResults) {
+      return Promise.resolve(prevResults.filter(node =>
+        [...node.edgesIn].some(edge => edge.prod)))
+    },
```

Comment:
```suggestion
    async '.prod' (prevResults) {
      return prevResults.filter(node =>
        [...node.edgesIn].some(edge => edge.prod))
    },
```
same on all the others, then it can't ever throw a sync exception

Reply:
Yep I'll audit for anything returning a promise to be an async for just this reason.